src/effects/shaders/halftone.frag

```diff
@@ -0,0 +1,205 @@
+#define SQRT2_MINUS_ONE 0.41421356
+#define SQRT2_HALF_MINUS_ONE 0.20710678
+#define SHAPE_DOT 1
+#define SHAPE_ELLIPSE 2
+#define SHAPE_LINE 3
+#define SHAPE_SQUARE 4
+
+uniform float radius;
+uniform float rotateR;
+uniform float rotateG;
+uniform float rotateB;
+uniform float scatter;
+uniform int shape;
+
+const int samples = 8;
+
+float hypot( float x, float y ) {
+
+	// vector magnitude
+	return sqrt( x * x + y * y );
+
+}
+
+float distanceToDotRadius( float channel, vec2 coord, vec2 normal, vec2 p, float angle, float rad_max ) {
+
+	// apply shape-specific transforms
+		float dist = hypot( coord.x - p.x, coord.y - p.y );
+		float rad = channel;
+
+		if ( shape == SHAPE_DOT ) {
```

Comment:
This dynamic `if-else` could be replaced with a preprocessor `if-else`. But you don't have to change it - I'll experiment with that later.

Reply:
Out of curiosity, how would you do this? would you use this.input.defines to set a definition, and then use the preprocessor if else in the shader?

Reply:
Exactly, maintain `SHAPE` as a macro and use `#if SHAPE == SHAPE_DOT...`.

Reply:
Cool, think I got it with f56d49a77e7928d777599cf0383c69156ec27d89 :)